### ISSUE TITLE
setup:upgrade action: adding trigger (mview.xml) to not exist table

### ISSUE THREAD

**Issue by chochold** [magento/magento2#34668]
### Preconditions (*)
1. Magento 2.4.3 - upgrade from 2.2.4 to 2.4.3
2. php 7.4
3. "amasty/sorting" - upgrade from "^2.6" to "^2.9" version

### Steps to reproduce (*)
1. run composer update
2. run bin/magento setup:upgrade

### Expected result (*)
1. upgraded Magento and Amaty module to newest version

### Actual result (*)

> The directory '/app/generated/metadata/' doesn't exist - skipping cleanup
Updating modules:
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'amasty_sorting_reviews_link' doesn't exist, query was: CREATE TRIGGER trg_amasty_sorting_reviews_link_after_insert AFTER INSERT ON amasty_sorting_reviews_link FOR EACH ROW
BEGIN
INSERT IGNORE INTO `catalogsearch_fulltext_cl` (`entity_id`) VALUES (NEW.`product_id`);
END

### Description (*)
In `vendor/amasty/sorting/etc/mview.xml` 

> <view id="catalogsearch_fulltext" class="\Magento\CatalogSearch\Model\Indexer\Mview\Action" group="indexer">
        <subscriptions>
            <table name="amasty_sorting_reviews_link" entity_column="product_id" />
        </subscriptions>
    </view>

Magento wanted to create trigger `trg_amasty_sorting_reviews_link_after_insert` but the table  `amasty_sorting_reviews_link` wasn't created yet, as you can see in the error .

Table `amasty_sorting_reviews_link` is created in the same Amasty module `vendor/amasty/sorting/etc/db_schema.xml`.

The problem is in the `setup/src/Magento/Setup/Console/Command/UpgradeCommand.php` file in `execute` function.

> $installer->removeUnusedTriggers();
$installer->installSchema($request);

First, magento removes all triggers and creates new ones in `removeUnusedTriggers()`, but the `amasty_sorting_reviews_link` table is created in the following `installSchema()` function.

Therefore we get an error there as I wrote above.
It might help to change the call position of these functions.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-November** [comment]
Hi @chochold ,
Amasty is a 3rd party extension and code of this extension is not part of https://github.com/magento/magento2 git repository. We are closing this issue as community is not able to provide fix for it in this repository. All questions, issue reports and fix for them should be addressed to the corresponding extension owners(support) on the [Magento Market place](https://marketplace.magento.com/extensions.html) page
Thank you.

**Comment by chochold** [comment]
Hi @engcom-November,

it is no 3rd party module problem. The main problem is in the Magento core setup:upgrade code. Amasty we used as an example.

The problem is in this file
`setup/src/Magento/Setup/Console/Command/UpgradeCommand.php`

> $installer->removeUnusedTriggers();
> $installer->installSchema($request);

Because, if you want to create a <subscription> on xxxxx table 

```
<view id="catalogsearch_fulltext"
 class="\Magento\CatalogSearch\Model\Indexer\Mview\Action"
 group="indexer">
         <subscriptions>
             <table name="xxxxx"
 entity_column="product_id" />
         </subscriptions>
</view>
```

And xxxxx table is created in the same module where we have `mview.xml`, setup:upgrade generates error because the table does not exist when triggers are generated.
Table will generate in the next command 
> $installer->installSchema($request);

**Comment by alek-s-andr** [comment]
We have faced the same issue. We created a new module, created **db_schema.xml** file which describes new table, also we created **mview.xml** file, but when we deploy this module to new environment (which don't has this table) we are getting the error.

![2022-12-02-13-24-03](https://user-images.githubusercontent.com/34573954/205229322-af003bb4-b665-4185-ac1c-54fb48ad71d8.png)

![2022-12-02-13-29-59](https://user-images.githubusercontent.com/34573954/205229922-c47935d7-b5f5-406e-bd6d-93f625ab9f2a.png)

**Comment by mamsincl** [comment]
**not limited to 3rd party modules**

![image](https://github.com/magento/magento2/assets/17334804/ed0a134f-3c61-4299-884d-5ca56d9d09d4)

module disabled by config.xml then re-enabled
cannot use CLI command to enable module as per CI/CD

**Comment by tuyennn** [comment]
@hostep  @mamsincl do you guys have any idea to handle this problem, this is bad situation has happened to us/

**Comment by tuyennn** [comment]
@engcom-November kindly reopen this issue, **this is not 3rd party module**, while Magento\Framework\Mview\Config\Reader was running to validate before the table was created by schema. In this case mview should ignore or run after the table creation.

**Comment by hostep** [comment]
@tuyennn, you can probably workaround the problem by temporarily setting indexers to 'realtime' mode, deploy your module, and reverting indexers to 'scheduled' mode.

This issue also reminds me of https://github.com/magento/magento2/pull/33092 which got included in Magento 2.4.7, are you on that version or a lower one?

**Comment by tuyennn** [comment]
@hostep good to know it fixed, well a patch is necessary while I'm on 2.4.5-p9
